CEL: allow multiple messageExpression strings to be returned

alexzielenski opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
Blocker for Declarative Validation for Native Types. Allows a CEL `list(string)` to be returned for `messageExpression`. Uses a CEL library to gate the feature, so it acquires ratcheting properties of other CEL libraries.

/cc @cici37 @jpbetz 

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Allows for CEL `messageExpression` under `x-kubernetes-validations` to accept a `list(string)` in addition to `string` to allow multiple errors to be constructed and returned. New CRDs will be able to use this change starting in 1.32.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

alexzielenski committed:
allow multiple messageExpression strings to be returned

cici37 commented:
/triage accepted

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -597,58 +599,85 @@ func fieldErrorForReason(fldPath *field.Path, value interface{}, detail string,
 // evalMessageExpression evaluates the given message expression and returns the evaluated string form and the remaining budget, or an error if one
 // occurred during evaluation.
-func evalMessageExpression(ctx context.Context, expr celgo.Program, exprSrc string, activation interpreter.Activation, remainingBudget int64) (string, int64, *cel.Error) {
+func evalMessageExpression(ctx context.Context, expr celgo.Program, exprSrc string, activation interpreter.Activation, remainingBudget int64) ([]string, int64, *cel.Error) {
```
Do we also need to pass in a `allowsMultipleMessageExpressions` bool to this function to reject list responses for VAP since we don't type check the expressions and hence don't know for sure what the response type will be until runtime?

alexzielenski replied:
This function is only called from the CRD Validation Rules CEL Validator, so that may not be necessary.

jpbetz replied:
Are we also adding mutliple message expression support to VAP or other places CEL is supported?  Should we?

alexzielenski replied:
We can do that for parity. I can follow-up later on. It would be important to do this release if we want to use the same CEL library trick, I think. Otherwise we'd need a separate flag for VAP

alexzielenski commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation_test.go:
```diff
@@ -9748,6 +9748,35 @@ func TestValidateCustomResourceDefinitionValidation(t *testing.T) {
 				invalid("spec.validation.openAPIV3Schema.properties[f@2].x-kubernetes-validations[0].rule"),
 			},
 		},
+		{
```
Added

alexzielenski commented on staging/src/k8s.io/apiserver/pkg/cel/library/multiple_message_expression.go:
```diff
@@ -0,0 +1,25 @@
+
+var multipleMessageExpressionLib = &multipleMessageExpression{}
+
+var MultipleMessageExpressionName = "MultipleMessageExpression"
```
Changed

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -597,58 +599,85 @@ func fieldErrorForReason(fldPath *field.Path, value interface{}, detail string,
+				return nil, remainingBudget - int64(*rtCost), &cel.Error{
+					Detail: "messageExpression failed to evaluate to a list of strings",
+				}
+			}
```
Can we add enough test cases to isolate the existing behavior?   I suspect the IsEquivalentType behavior might be lenient.

(Also, when adding VAP support in a future PR, Having a test that fails for VAP (which is runtime only) if a list of non-strings is returned seems valuable)

alexzielenski replied:
I added a test `TestOutputType` to probe this behavior. It looks like the issue isn't with `IsEquivalentType` - it seems to just be that `ref.Val.Value()` return value is an implementation detail. If you want guarantees about the runtime Go value of a `ref.Val` you need to use `ref.Val.ConvertToNative` instead:


```go
	// Value returns the raw value of the instance which may not be directly compatible with the expression
	// language types.
	Value() any

	// ConvertToNative converts the Value to a native Go struct according to the
	// reflected type description, or error if the conversion is not feasible.
	//
	// The ConvertToNative method is intended to be used to support conversion between CEL types
	// and native types during object creation expressions or by clients who need to adapt the,
	// returned CEL value into an equivalent Go value instance.
	//
	// When implementing or using ConvertToNative, the following guidelines apply:
	// - Use ConvertToNative when marshalling CEL evaluation results to native types.
	// - Do not use ConvertToNative within CEL extension functions.
	// - Document whether your implementation supports non-CEL field types, such as Go or Protobuf.
	ConvertToNative(typeDesc reflect.Type) (any, error)
```

alexzielenski replied:
I changed our message expression evaluation implementation to use `ConvertToNative` rather than Value in response to this

alexzielenski committed:
add test to probe behavior of IsEquivalentType, ref.Val.Value, and ast.OutputType

Value seems to be an implementation detail, and convertToNative is recommended approach

jpbetz commented:
/approve
/lgtm

Thanks @alexzielenski !

liggitt commented:
> Blocker for Declarative Validation for Native Types

I don't understand why we're changing the CRD API for this... is that what's happening in this PR?

alexzielenski commented:
> Blocker for Declarative Validation for Native Types
> I don't understand why we're changing the CRD API for this... is that what's happening in this PR?

Multiple error message support is something that improves CRD API even if native type validation doesn't pan out - we were also considering adding support to VAP. We can scope the change to be under a flag only turned on for native types but I would consider that unfortunate.

liggitt commented:
> > Blocker for Declarative Validation for Native Types
> > I don't understand why we're changing the CRD API for this... is that what's happening in this PR?
> 
> Multiple error message support is something that improves CRD API even if native type validation doesn't pan out - we were also considering adding support to VAP. We can scope the change to be under a flag only turned on for native types but I would consider that unfortunate.

Hrmm... every time we change the CRD CEL API surface we fragment the Kubernetes versions a given CRD can support, so we shouldn't do that lightly.

alexzielenski commented:
> Hrmm... every time we change the CRD CEL API surface we fragment the Kubernetes versions a given CRD can support, so we shouldn't do that lightly.

It makes sense that we wouldn't want to have overly complicated rules for what CEL features you can and can't use to target a particular k8s version. I still think this makes CRD Validation Rules better and would like to see it supported. What if we would soak it for longer than 1 release - is there an acceptable number of releases to soak a CEL change like this? For ex. I see gateway API supports at least the 5 most recent k8s releases.

liggitt commented:
I think we should only fragment CRD version support for clear capability gains ... things you *couldn't* do before the new capability enables. I have a hard time seeing this in that category... do you have an example of a CEL messageExpression this would enable and what the alternative today would be without this support?

alexzielenski commented:
For example we have name formats library in CEL used like: `format.dns1123Subdomain().validate(myString)` returning `optional(list(string))`. Without this feature, the only way to ensure each returned error can appear in an individual error message is to write multiple rules like: 

```yaml
- rule: !format.dns1123Subdomain().validate(myString).hasValue()
  messageExpression: format.dns1123Subdomain().validate(myString).value()[0]
- rule: !format.dns1123Subdomain().validate(myString).optMap(x, x.size() > 1).hasValue()
  messageExpression: format.dns1123Subdomain().validate(myString).value()[1]
- rule: !format.dns1123Subdomain().validate(myString).optMap(x, x.size() > 2).hasValue()
  messageExpression: format.dns1123Subdomain().validate(myString).value()[2]
- rule: !format.dns1123Subdomain().validate(myString).optMap(x, x.size() > 3).hasValue()
  messageExpression: format.dns1123Subdomain().validate(myString).value()[3]
```

A workaround could be to use `join` for worse UX (individual errors can be long, so likely only first is read)

```yaml
- rule: !format.dns1123Subdomain().validate(myString).hasValue()
  messageExpression: format.dns1123Subdomain().validate(myString).value().join("\n")
```

I'm fine gating this for only native types if the workaround is seen as sufficient for CRDs
